subprojects/composite-builds/src/main/java/org/gradle/composite/internal/IncludedBuildArtifactBuilder.java

```diff
@@ -16,56 +16,40 @@
 
 package org.gradle.composite.internal;
 
-import com.google.common.collect.LinkedHashMultimap;
-import com.google.common.collect.Multimap;
 import org.gradle.api.GradleException;
 import org.gradle.api.artifacts.component.BuildIdentifier;
-import org.gradle.initialization.IncludedBuildExecuter;
+import org.gradle.initialization.IncludedBuildTaskGraph;
 import org.gradle.initialization.ReportedException;
 import org.gradle.internal.component.model.ComponentArtifactMetadata;
 import org.gradle.internal.exceptions.Contextual;
 import org.gradle.internal.exceptions.LocationAwareException;
 
-import java.util.Collection;
-
 class IncludedBuildArtifactBuilder {
-    private final Multimap<BuildIdentifier, String> tasksForBuild = LinkedHashMultimap.create();
-    private final IncludedBuildExecuter includedBuildExecuter;
-
-    IncludedBuildArtifactBuilder(IncludedBuildExecuter includedBuildExecuter) {
-        this.includedBuildExecuter = includedBuildExecuter;
-    }
+    private final IncludedBuildTaskGraph includedBuildTaskGraph;
 
-    public void willBuild(ComponentArtifactMetadata artifact) {
-        if (artifact instanceof CompositeProjectComponentArtifactMetadata) {
-            CompositeProjectComponentArtifactMetadata compositeBuildArtifact = (CompositeProjectComponentArtifactMetadata) artifact;
-            BuildIdentifier buildId = getBuildIdentifier(compositeBuildArtifact);
-            addTasksForBuild(buildId, compositeBuildArtifact);
-        }
+    IncludedBuildArtifactBuilder(IncludedBuildTaskGraph includedBuildTaskGraph) {
+        this.includedBuildTaskGraph = includedBuildTaskGraph;
     }
 
     public void build(BuildIdentifier requestingBuild, ComponentArtifactMetadata artifact) {
         if (artifact instanceof CompositeProjectComponentArtifactMetadata) {
             CompositeProjectComponentArtifactMetadata compositeBuildArtifact = (CompositeProjectComponentArtifactMetadata) artifact;
-            BuildIdentifier buildId = getBuildIdentifier(compositeBuildArtifact);
-            Collection<String> tasksToExecute = addTasksForBuild(buildId, compositeBuildArtifact);
-            execute(requestingBuild, buildId, tasksToExecute);
+            BuildIdentifier targetBuild = getBuildIdentifier(compositeBuildArtifact);
+            for (String taskName : compositeBuildArtifact.getTasks()) {
+                includedBuildTaskGraph.addTask(requestingBuild, targetBuild, taskName);
+            }
+            execute(targetBuild);
```

Comment:
This is correct: when more than 1 task is registered for an included build we attempt to execute them in a single invocation, to avoid the overhead of multiple build invocations.

When you say:
> it would appear as if the first task caused by A and B to be built.
Do you mean in the console? In Build Operations?
This is very likely to change when we execute included builds in parallel: the dependent task will wait for the included build task to complete, but it will not directly "cause" the task to execute.